add TorchSharp replace

toolgood opened the pull request.

toolgood committed:
add TorchSharp replace

GeorgeS2019 commented:
@toolgood  Good Job!

uxmal commented:
This is a very nice contribution, but I think it might benefit from rearchitecting. You are doing post processing of the files that pytocs is generating, but it seems to me that it would be more efficient to post process while still inside the pytocs main process. If you're willing to have some patience, would you consider a redesign of this? Here is an idea:

Pytocs provides the following interface:
```C#
public interface IPostProcessor {
     public CodeCompileUnit PostProcess(CodeCompileUnit unit);
}
```

On the command line, introduce a `--post-process` flag which takes 1 or more arguments. Invoking pytocs with the flag could look something like:
```
pytocs --post TorchSharp,Math,Os -r .
```
Any implementation of `IPostProcessor` can then examine the `CodeCompileUnit` and where needed, mutate it.

I will add `IPostProcessor` and an example of how it could be used to adjust namespaces like your PR does. Would you be agreeable to change the code to follow the example?

uxmal commented:
After taking a look to see what changes would be needed, and the complexity of implementing such an `IPostProcessor` I'm thinking maybe it's overengineering. I think it'd be better to proceed with post processor as a separate tool, like in this PR.

There are some comments I'd like to make before I can accept the PR. I will have time later today to add those.

uxmal commented on src/TorchCs/TorchCs.csproj:
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
```
Please change to net6.0

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,391 @@
+
+            return text;
+        }
+        #region replaceNamespace
```
The `#region` directives here adds only clutter. Please remove it and all other `#region` directives in this file.

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,391 @@
+            return text;
+        }
+        #region replaceNamespace
+        static string replaceNamespace(string text)
```
Please always state explictly the access modifiers, be it `private`, `protected`, `internal`, or `public`.

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,391 @@
+using System;
```
Don't forget to add a license for your contribution here. You can model it after the other license headers in the Pytocs project.

uxmal commented on src/pytocs.sln:
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
```
Once the PR is accepted I will take care of restructuring the file system so that it gets placed in an appropriate place in the VS Solution Explorer.

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,391 @@
+        static string replaceConstructor(string text)
+        {
+            var ms = Regex.Matches(text, @"public class (\S+)[\s \t]*: nn.Module");
+            if (ms.Count > 0)
```
The check for `ms.Count > 0` is not necessary. The `foreach` below won't do anything if the `ms` collection is empty.

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,391 @@
+                var ps = methodInfo.GetParameters();
+                foreach (var p in ps)
+                {
+                    for (int i = 0; i < 5; i++)
```
What is this magic `5` number here? Please provide a symbolic constant for it.

uxmal reviewed (commented):
I've added some comments that need to be addressed. Ideally, I'd like the new project hosted in the directory `src/Extensions/TorchCs` as it is possible other extensions might be necessary.

GeorgeS2019 commented:
@uxmal  thnx for your support. I think @toolgood is on holiday NOW :-)

There are now two tentative use cases.
- take a pytoch code and use this PR to improve the quality of the port to TorchSharp.
- It is WIP.  It seems ChatGPT could attempt to do the port (obviously with lots of errors), 

>Preliminary, we believe the AI simply creates a Pytorch like syntax but with e.g. c# `using, namespace `etc. Need to investigate further.  Perhaps there is something to learn from ChatGPT :-)

toolgood committed:
Merge branch 'uxmal:master' into master

GeorgeS2019 commented:
@toolgood  is back :-)

Commit:
修改 代码

toolgood commented:
The code used is as follows 
``` cs
            var folder = "F:\\BaiduNetdiskDownload\\TestDR\\TestDR";
            TorchUtil.ReplaceFolder(folder);
            TorchUtil.CreateNetstandardCode(folder);
```

GeorgeS2019 commented:
@toolgood  thx!

@uxmal  Including you too to discuss the feasibility to make the folder, not a pytorch project folder BUT pytorch jupyter notebook.

Jupyter notebook are text files with e.g. pytorch codes.

Ideally, the suggestion is to take such e.g. a folder of pytorch notebooks (typically e.g. a pytorch tutorial or pytorch 101) and add in the additional notebook header configuration to make the notebooks recognized by e.g. .NET interactive to use c# kernel.

Then add the cell "package install"  TorchSharp 

Then work on each cell top down, ignoring the markdown cells.

I believe such pytorch code structure in jupyter are MORE STRUCTURE then the typical e.g a folder of pytorch codes.


## WHY?

@uxmal  For John, who has been dedicated so many years in this project, there is NOW a valid use case to make an impact to e.g. Universities or Colleges teaching pytorch, instead in python but in C# 

If this works, I am confident the .NET interactive team could consider integrating this as part of the PolyGlot Notebook pitch.

@toolgood  thx for making this idea ONE step closer through your initiative!

uxmal commented on src/TorchCs/Resources/netstandard.cs:
```diff
@@ -0,0 +1,205 @@
+                return path1 + (firstHasSeparator ? "" : DirectorySeparatorCharAsString) + path2 + (secondHasSeparator ? "" : DirectorySeparatorCharAsString) + path3;
+            }
+
+            public static string join(string path1, string path2, string path3, string path4)
```
You need to simplify this into a single `join(params string[] paths)`:
```C#
public static string join(params string[] paths) {
    return System.IO.Path.Join(paths);
}
```

uxmal commented on src/TorchCs/TorchUtil.cs:
```diff
@@ -0,0 +1,481 @@
+            return text;
+        }
+
+        private static string replaceListSlice(string text)
```
Please add a comment to give an example of what "replaceListSlice" does.

uxmal commented on src/TorchCs/Resources/netstandard.cs:
```diff
@@ -0,0 +1,205 @@
+
+            public static bool exists(string path)
+            {
+                if (File.Exists(path)) {
```
This can be simplified:
```C#
public static bool exists(string path) {
    return (File.Exists(path) || Directory.Exists(path);
}
```

uxmal commented on src/TorchCs/Resources/netstandard.cs:
```diff
@@ -0,0 +1,205 @@
+            return dict.Keys;
+        }
+
+        public static (long, long) ToLong2(this long[] array)
```
Please add a comment explaining what these are used for.

uxmal commented on src/TorchCs/Resources/netstandard.cs:
```diff
@@ -0,0 +1,205 @@
+                return hasSeparator ? string.Concat(path1, path2) : string.Concat(path1, DirectorySeparatorCharAsString, path2);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Not sure why you're copying the DirectorySeparatorChar from `System.IO.Path` here.

uxmal reviewed (commented):
A few more fixes. Also, are you going to add tests to this new code? It would be a shame if changes in pytocs cause all this code to break.

A gentle appeal: please write all code / comments / commit comments in English. It's not my native language either, but programmers from all over the world will be most familiar with English.

Commit:
Add comments and optimize the code.

Commit:
New conversion.
example:
   public object in_channels; => public int in_channels;
   (B, T) = x.size()  => var (B, T) = x.size().ToLong2();
   (B, T, N) = x.size()  => var (B, T, N) = x.size().ToLong3();
   (B, T, N,_) = x.size()  => var (B, T, N,_) = x.size().ToLong4();

uxmal commented:
Looks good. I will refactor this as mentioned above.
